### DISCUSSION TITLE
레플리카셋에 대하여 질문드립니다.

### DISCUSSION THREAD

**Discussion by sc303030** [yoonjeong-kwon/fastcampus-kubernetes#29] [Q&A] [answered]
안녕하세요. 
**Ch.3 Pod 복제와 배포에 날개달기 - 01. ReplicaSet** 에 대하여 궁금한 점이 있습니다.
레플리카셋을 3으로 지정하고 pod를 생성하면 3개의 파드가 생기는데
1. 그럼 3개의 파드(1번, 2번, 3번 파드로 지칭)중 1번 파드만 일을 하나요?
    1-1. 나머지 2개는 1번이 망가질 경우를 대비해서 휴식상태에 있는건지 궁금합니다.
만약 그게 아니라면 레플리카셋(1번, 2번, 3번 파드)가 들어오는 트래픽을 분산하여 처리한다는 것인데
2. 레플리카셋은 로드 밸런싱 기능이 없어서 로드 밸런싱이 있어야먄 트래픽 분산 처리가 가능한지, 아니면 본래의 기능중 트래픽 분산 처리 기능이 있는지 질문드려봅니다.

감사합니다.

**Comment by LEEWONHOI** [comment]
@sc303030 

안녕하세요. 강사님을 대신해 제가 알고있는 내용을 답변드리겠습니다. 제 답변은 참고만 해주세요.

1. ReplicaSet의 기본 개념 및 동작 방식:
R**eplicaSet의 주요 목적은 지정된 수의 파드 복제본을 유지하는 것입니다.** 즉, ReplicaSet을 3으로 설정하면 Kubernetes는 3개의 파드 인스턴스를 항상 실행 상태로 유지하려고 합니다. 이는 파드 중 하나가 실패하거나 삭제될 경우, ReplicaSet이 자동으로 다른 파드를 생성하여 원하는 수의 파드가 유지되도록 보장합니다.

1-1. 파드의 작업 분담:
ReplicaSet이 생성한 모든 파드는 각각 독립적으로 작동합니다. **이는 모든 파드가 병렬로 작업을 처리할 수 있으며, 특정 파드만 작업을 처리하고 나머지가 대기 상태에 있는 것이 아닙니다.** 각 파드는 동일한 컨테이너 이미지와 설정을 기반으로 하여 동일한 작업 또는 서비스를 제공할 수 있습니다.

2. 로드 밸런싱과 트래픽 분산:
ReplicaSet 자체는 로드 밸런싱 기능을 내장하고 있지 않습니다. **ReplicaSet은 단순히 지정된 수의 파드를 유지하는 역할을 합니다. 로드 밸런싱은 트래픽을 여러 파드에 분산시키는 별도의 메커니즘을 통해 처리됩니다.** Kubernetes에서는 일반적으로 Service 리소스를 사용하여 로드 밸런싱을 구현합니다. Service는 클러스터 내부 또는 외부의 트래픽을 여러 파드로 분산시키는 역할을 하며, 각 파드에 고르게 트래픽을 분배할 수 있습니다.

결론적으로, 
1. **ReplicaSet은 파드의 복제본을 유지하고 관리하는 데 집중합니다.** 별도의 트래픽 분산은 고려되지 않습니다.
2. 질문주신 **로드 밸런싱과 트래픽 분산은 주로 Service를 통해 처리됩니다.** 모든 파드는 ReplicaSet으로 인해 활성 상태로 있으며 필요에 따라 트래픽을 처리할 수 있습니다.

이미 잘 아시겠지만, 쿠버네티스 네트워크 정책은 아래의 공식 문서에서 더 자세히 확인할 수 있습니다. 참고해주세요.
https://kubernetes.io/ko/docs/concepts/services-networking/

**Comment by yoonjeong-kwon** [comment, marked as answer]
> 안녕하세요. **Ch.3 Pod 복제와 배포에 날개달기 - 01. ReplicaSet** 에 대하여 궁금한 점이 있습니다. 레플리카셋을 3으로 지정하고 pod를 생성하면 3개의 파드가 생기는데
> 
> 1. 그럼 3개의 파드(1번, 2번, 3번 파드로 지칭)중 1번 파드만 일을 하나요?
>    1-1. 나머지 2개는 1번이 망가질 경우를 대비해서 휴식상태에 있는건지 궁금합니다.
>    만약 그게 아니라면 레플리카셋(1번, 2번, 3번 파드)가 들어오는 트래픽을 분산하여 처리한다는 것인데
> 2. 레플리카셋은 로드 밸런싱 기능이 없어서 로드 밸런싱이 있어야먄 트래픽 분산 처리가 가능한지, 아니면 본래의 기능중 트래픽 분산 처리 기능이 있는지 질문드려봅니다.
> 
> 감사합니다.

@sc303030 님, 안녕하세요.
너무 늦은 답변 죄송하다는 말씀드리며 이후 강의에서 궁금증을 해결하셨을지 궁금하네요.

1. ReplicaSet을 통해 3개의 파드가 순차적으로 생성됩니다. 만약 파드의 모든 컨테이너가 정상적으로 실행되고 있다면 파드 상태는 Running으로 전환됩니다. 즉 3개 모두 요청을 받을 수 있는 상태입니다.

2. 쿠버네티스에서 로드밸런싱을 담당하는 오브젝트는 Service 입니다. Service는 클러스터 내외부 트래픽을 받아 Service가 관리하는 파드 간에 트래픽을 분산시키는 역할을 합니다. 로드밸런싱 방식은 ClusterIP, NodePort, LoadBalancer 설정에 따라 다릅니다.

감사합니다.
